Escaped backslashes in the bib file cause the building process to crash

# Problem

A bib file with `url=https://doi.org/10.1007/978-3-031-56940-1\_2` causes the build process to crash, since it cannot handle the backslash.

```
> dice-website@1.1.0 build /home/runner/work/dice-website/dice-website/scripts
> gatsby build

success open and validate gatsby-configs - 0.041s
success load plugins - 0.940s
success onPreInit - 0.029s
success delete html and css files from previous builds - 0.002s
success initialize cache - 0.004s
success copy gatsby files - 0.041s
success onPreBootstrap - 0.011s
success createSchemaCustomization - 0.010s
error Error parsing papers/2df4da89aaa2bd628ec730bf3f1abef3c_dice-research.ttl! Error: Unexpected "<[https://doi.org/10.1007/978-3-031-56940-1\\_2](https://doi.org/10.1007/978-3-031-56940-1//_2)>;" on line 13.
    at N3Lexer._syntaxError (/home/runner/work/dice-website/dice-website/scripts/node_modules/n3/lib/N3Lexer.js:466:17)
    at reportSyntaxError (/home/runner/work/dice-website/dice-website/scripts/node_modules/n3/lib/N3Lexer.js:394:21)
    at N3Lexer._tokenizeToEnd (/home/runner/work/dice-website/dice-website/scripts/node_modules/n3/lib/N3Lexer.js:375:86)
    at /home/runner/work/dice-website/dice-website/scripts/node_modules/n3/lib/N3Lexer.js:484:83
    at internal/process/task_queues.js:141:7
    at AsyncResource.runInAsyncScope (async_hooks.js:197:9)
    at AsyncResource.runMicrotask (internal/process/task_queues.js:138:8) {
  context: {
    token: undefined,
    line: 13,
    previousToken: { line: 13, type: 'prefixed', value: 'url', prefix: 'schema' }
  }
}
error UNHANDLED REJECTION Unexpected "<[https://doi.org/10.1007/978-3-031-56940-1\\_2](https://doi.org/10.1007/978-3-031-56940-1//_2)>;" on line 13.


  Error: Unexpected "<
[https://doi.org/10.1007/978-3-031-56940-1\\_2](https://doi.org/10.1007/978-3-031-56940-1//_2)
>;" on line 1  3.
  
  - N3Lexer.js:466 N3Lexer._syntaxError
    [scripts]/[n3]/lib/N3Lexer.js:466:17
  
  - N3Lexer.js:394 reportSyntaxError
    [scripts]/[n3]/lib/N3Lexer.js:394:21
  
  - N3Lexer.js:375 N3Lexer._tokenizeToEnd
    [scripts]/[n3]/lib/N3Lexer.js:375:86
  
  - N3Lexer.js:484 
    [scripts]/[n3]/lib/N3Lexer.js:484:83
  
  - task_queues.js:141 
    internal/process/task_queues.js:141:7
  
  - async_hooks.js:197 AsyncResource.runInAsyncScope
    async_hooks.js:197:9
  
  - task_queues.js:138 AsyncResource.runMicrotask
    internal/process/task_queues.js:138:8
  

not finished source and transform nodes - 1.268s
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! dice-website@1.1.0 build: `gatsby build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the dice-website@1.1.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/runner/.npm/_logs/2025-05-19T12_41_22_057Z-debug.log
Error: Process completed with exit code 1.
```

# Solution

Replace all backslashes in URLs when transforming the bib data into RDF.

--- a/scripts/papers-to-ttl/index.js
+++ b/scripts/papers-to-ttl/index.js
@@ -73,13 +73,38 @@ function checkUrl(url) {
 }
 
 /**
- * A simple method for pre processing PDF URLs removing '\\_' and '\\%'.
+ * Extract a raw URL from possible Markdown like: [text](https://example)
  */
-function preprocessPdfUrl(url) {
-  if (url === null || url === undefined) {
-    return '';
+function extractUrl(s) {
+  if (!s) return '';
+  const str = String(s);
+  const m = str.match(/\((https?:\/\/[^)]+)\)/); // [text](URL)
+  if (m) return m[1];
+  const bare = str.match(/https?:\/\/[^\s)>]+/); // fallback: first http(s)://...
+  return bare ? bare[0] : str;
+}
+
+// Normalize a URL:
+//  - remove backslashes used as escapes (\_ \%)
+//  - collapse accidental multiple slashes is path (keep scheme //)
+//  - trim
+function normalizeUrl(s) {
+  if (!s) return '';
+  let url = extractUrl(String(s).trim());
+  url = url.replace(/^<+|>+$/g, '');
+  url = url.replace(/\\/g, '');
+  const m = url.match(/^(https?:\/\/[^/]+)(\/.*)?$/i);
+  if (m) {
+    const host = m[1];
+    const path = (m[2] || '').replace(/\/{2,}/g, '/');
+    url = host + path;
   }
-  return url.replace(/\\/g, '');
+  return url;
+}
+
+// Keep legacy behaivor for PDF URLs but reuse the generic normalizer.
+function preprocessPdfUrl(url) {
+  return normalizeUrl(url);
 }
 
 /**
@@ -143,10 +168,23 @@ const main = async () => {
         writeLiteral('tag', tag);
       });
     }
-    // make sure that URL is valid before adding it
-    if (checkUrl(paper.url)) {
-      writeUrl(`${prefixes.schema}url`, paper.url);
+
+    // sanitize and add the canonical URL
+    const cleanUrl = normalizeUrl(paper.url);
+    if (checkUrl(cleanUrl)) {
+      writeUrl(`${prefixes.schema}url`, cleanUrl);
+    } else if (paper.doi) {
+      const doiUrl = normalizeUrl(
+        `https://doi.org/${String(paper.doi).replace(
+          /^https?:\/\/(dx\.)?doi\.org\//i,
+          ''
+        )}`
+      );
+      if (checkUrl(doiUrl)) {
+        writeUrl(`${prefixes.schema}url`, doiUrl);
+      }
     }
+
     writeUrl(`${prefixes.schema}bibsonomyId`, paper.id);
     var pdfUrl = preprocessPdfUrl(paper['bdsk-url-1'] || paper['1']);
     if (checkUrl(pdfUrl)) {

